src/runtimes/index.js

```diff
@@ -1,8 +1,82 @@
+const { extname, basename } = require('path')
+
 const goRuntime = require('./go')
 const jsRuntime = require('./node')
 const rustRuntime = require('./rust')
 
+/**
+ * Finds functions for a list of paths using a specific runtime. The return
+ * value is an object containing an array of the functions found (`functions`)
+ * and an array with the paths that haven't been recognized by the runtime
+ * (`remainingPaths`).
+ *
+ * @param   {Array<String>} paths
+ * @param   {Object} runtime
+ * @returns {Promise<Object>}
+ */
+const findFunctionsInRuntime = async function ({ dedupe = false, paths, runtime }) {
+  const functions = await runtime.findFunctionsInPaths(paths)
+
+  // If `dedupe` is true, we use the function name (`filename`) as the map key,
+  // so that `function-1.js` will overwrite `function-1.go`. Otherwise, we use
+  // `srcPath`, so that both functions are returned.
+  const key = dedupe ? 'name' : 'srcPath'
+
+  // Augmenting the function objects with additional information.
+  const augmentedFunctions = functions.map((func) => [
+    func[key],
+    {
+      ...func,
+      extension: extname(func.mainFile),
+      filename: basename(func.srcPath),
+      runtime,
+    },
+  ])
+  const usedPaths = new Set(augmentedFunctions.map(([path]) => path))
```

Comment:
Would it make sense to use the following line instead?

```js
  const usedPaths = new Set(functions.map((func) => func[key]))
```

This would allow `argumentedFunctions` to be directly an array of objects:

```js
const augmentedFunctions = functions.map((func) => ({
  ...func,
  extension: extname(func.mainFile),
  filename: basename(func.srcPath),
  runtime,
}))
```

This, in turn, would remove the need to use a `Map()` in `getFunctionsFromPaths()`, returning an array instead. This would remove the need for `map.values()` and iterator methods in consumers.

What do you think?

Reply:
`augmentedFunctions` is returning an array where the first element is the function key and the second element is the function object. This syntax is compatible with [`Map`'s constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/Map#creating_a_new_map), so that by the time we've cycled through all run times and we construct a new `Map` at the end of `getFunctionsFromPaths`, we've de-duped any functions that share the same name.

If `augmentedFunctions` returns an array of objects, this de-duping would not be possible, unless I misunderstood your suggestion.

Reply:
Yes, you're right, my suggestion would not work :+1: